src/tox_travis.py

```diff
@@ -99,13 +101,104 @@ def get_default_envlist(version):
     return split_env(travis_section.get(version, default_envlist))
 
 
+def get_desired_envs(config, version):
+    """Get the expanded list of desired envs."""
+    if 'tox:travis' in config.sections:
+        return _legacy_get_desired_envs(config, version)
+
+    # Parse the travis section
+    travis_section = config.sections.get('travis', {})
+    default_envlist = get_default_envlist(version)
+    default_envlist = split_env(default_envlist)
+
+    # Get python version factors
+    python_factors = parse_dict(travis_section.get('python', ''))
+    python_factors = python_factors.get(version, '')
+    python_factors = split_env(python_factors)
+
+    # Get os version factors
+    os_factors = parse_dict(travis_section.get('os', ''))
+    os_factors = os_factors.get(os.environ.get('TRAVIS_OS_NAME'), '')
+    os_factors = split_env(os_factors)
+
+    # Combine python & os version factors
+    desired_factors = [
+        (python_factors + os_factors) or default_envlist
+    ]
+
+    # Parse the environment factors
+    env_section = config.sections.get('travis:env', {})
+
+    # Use the travis section and environment variables
+    # to determine the desired tox factors.
+    desired_factors += [
+        get_env_factors(env_section, envvar)
+        for envvar in env_section
+    ]
+
+    # filter empty factors and join
+    return reduce_factors(filter(None, desired_factors))
+
+
+def get_env_factors(env_section, envvar):
+    """Derive a list of tox factors from the travis:env section
+    using the current environment variables. For example, given
+    the following travis:env section:
+
+        [travis:env]
+        DJANGO =
+            1.9: django19
+            1.10: django110
+
+    If the current environment defines DJANGO as '1.9', then getting the
+    DJANGO envvar would return a ['django19'] factor list.
+
+    """
+    env_factors = parse_dict(env_section.get(envvar, ''))
+    return env_factors.get(os.environ.get(envvar))
+
+
+def parse_dict(value):
+    """Parse a dict value from the tox config. Values support comma
+    separation and brace expansion. ex:
+
+        python =
+            2.7: py27, docs
+            3: py{35,36}
+
+        >>> parse_dict("2.7: py27, docs\n3: py{35,36}")
+        {'2.7': ['py27', 'docs'], '3': ['py35', 'py36']}
+
+    """
+    # key-value pairs are split by line, may contain extraneous whitespace
+    kv_pairs = [line.strip() for line in value.splitlines()]
+    kv_pairs = [kv.strip().split(':', 1) for kv in kv_pairs if kv]
+
+    # values may be comma-separated
+    return dict(
+        (key.strip(), split_env(value))
+        for key, value in kv_pairs
+    )
+
+
+def reduce_factors(factors):
+    """Reduce sets of factors into a single list of envs.
+
+        >>> reduce_factors([['py27', 'py35'], ['django110']])
+        ['py27-django110', 'py35-django110']
+
+    """
+    envs = product(*factors)
+    envs = ['-'.join(env) for env in envs]
+    return envs
```

Comment:
Before I forget about it, duplicated factors need to be collapsed. For instance:

```pycon
>>> reduce_factors(['py27', 'py35'], ['py35'])
['py27-py35', 'py35']
```

The first is obviously non-sensical, But `py35-py35` would be redundant, but also clear that it's likely wanting to just run the `py35` environment.